### DISCUSSION TITLE
Use SVG directly as React component in common ui library

### DISCUSSION THREAD

**Discussion by agusterodin** [vercel/turborepo#638] [Help] [answered]
Typically in a Next or Create React App application to import an SVG image as a React Component I use:

```import { ReactComponent as ExampleIcon } from 'common-ui-library/images/icon.svg'```

Unfortunately since a common ui library (packages/ui in the starter boilerplate) needs to export things as proper javascript modules and doesn't have Webpack or other bundler associated with it I can't take advantage of SVGR.

How would you suggest to get this working? I would hate to need to set up an "images" package that uses Rollup or another bundler just to use SVG images in my React code ergonomically. 

Wouldn't that would entail the use of some sort of watcher that needs to run in a separate process? The last thing I want to need to do is manually rerun a tool like SVGO every time I add a new image while developing.

**Comment by erwol** [comment, marked as answer]
You can achieve this by adding a `types.d.ts` under your `common-ui-library` directory:
```
declare module '*.svg' {
  const content: any;
  export default content;
}
```
Then on your `common-ui-library/index.ts` do:
```
export { default as Icon } from './icon.svg';
```
And then on your app just do:
```
import { Icon  } from 'common-ui-library'; 
export const Img = () => <img src={Icon.src} />;
```
More info here: https://webpack.js.org/guides/typescript/#importing-other-assets

**Reply by agusterodin** [reply]
Awesome! 

Yeah I didn't realize that next-transpile-modules means that your common UI library will be built by the Next process (allowing you to take advantage of all the webpack plugins you have set up) and that no additional bundler process is required. CSS module loading and SVGR in my common UI library work like a charm. :D

**Reply by ArpitaGanatra** [reply]
does this let you display icons on server side, or will the icons load on client side?

if not, is there any other way than custom icon library? @Erwol

**Comment by SeanRoberts** [comment]
I'm not using Next so I needed something that could just export React components. I know you said you don't want to re-run a command every time you add an icon and unfortunately this doesn't solve that problem, but I don't add new icons that frequently so having to restart my dev task isn't a huge deal to me.  Hopefully this will help someone else:

1. Create a new package called `icons` 
2. Add this to package.json:

```json
{
  "name": "icons",
  "version": "0.0.0",
  "description": "Icons",
  "devDependencies": {
    "@svgr/cli": "^6.1.2"
  },
  "module": "src/index.ts",
  "scripts": {
    "svgr": "svgr --typescript --icon --title-prop -d src -- assets",
    "build": "yarn svgr",
    "dev": "yarn svgr"
  }
}
```

3. Load up all your svgs into `packages/icons/assets`
4. Add `"icons": "*"` to your dependencies of any app that needs icons.
5. `import { MyIcon } from 'icons'`

Currently the SVGR package doesn't have a --watch option, otherwise I think that might negate the need to restart the dev server when you add/change an icon.  It's probably also entirely possible to write a script using chokidar or something to re-run the svgr command when changes are detected in the assets directory.  

This is currently working for me in a Vite project without any additional configuration.  Other tools might need to be configured to transpile the modules from the icons package.  Hope this helps someone!
